packages/core/lib/patchers/aws3_p.ts

```diff
@@ -62,110 +82,110 @@ function addFlags(http: HttpResponse, subsegment: Subsegment, err?: SdkError): v
     }
+    return next(args);
+  }
 
-    const subsegment = segment.addNewSubsegment(serviceIdentifier);
-    subsegment.addAttribute('namespace', 'aws');
-
-    try {
-      const res = (await send.apply(client, [command])) as Output;
+  const subsegment: Subsegment = segment.addNewSubsegment(service);
+  subsegment.addAttribute('namespace', 'aws');
+  const parent = (segment instanceof Subsegment ? segment.segment : segment);
 
-      if (!res) throw new Error('Failed to get response from instrumented AWS Client.');
+  args.request.headers['X-Amzn-Trace-Id'] = stringify(
+    {
+      Root: parent.trace_id,
+      Parent: subsegment.id,
+      Sampled: parent.notTraced ? '0' : '1',
+    },
+    ';',
+  );
 
+  let res;
+  try {
+    res = await next(args);
+    if (!res) throw new Error('Failed to get response from instrumented AWS Client.');
+
+    // TODO: Figure out how to get region string. Should look like 'await options.region()'
+    const [aws, http] = await buildAttributesFromMetadata(
+      service,
+      operation,
+      await config.region(),
+      res,
+    );
+
+    subsegment.addAttribute('aws', aws);
+    subsegment.addAttribute('http', http);
+
+    addFlags(http, subsegment);
+    subsegment.close();
+    return res;
+  } catch (err) {
+    if (err.$metadata) {
       const [aws, http] = await buildAttributesFromMetadata(
-        client,
-        command,
-        res.$metadata,
+        service,
+        operation,
+        '',
+        err,
       );
 
       subsegment.addAttribute('aws', aws);
       subsegment.addAttribute('http', http);
-
-      addFlags(http, subsegment);
-      subsegment.close();
-      return res;
-    } catch (err) {
-      if (err.$metadata) {
-        const [aws, http] = await buildAttributesFromMetadata(
-          client,
-          command,
-          err.$metadata,
-        );
-
-        subsegment.addAttribute('aws', aws);
-        subsegment.addAttribute('http', http);
-        addFlags(http, subsegment, err);
-      }
-
-      const errObj = { message: err.message, name: err.name, stack: err.stack || new Error().stack };
-
-      subsegment.close(errObj, true);
-      throw err;
+      addFlags(http, subsegment, err);
     }
-  };
-
-  client.middlewareStack.add(
-    (next) => async (args: any) => {
-      const segment = manualSegment || contextUtils.resolveSegment();
-      if (!segment) return next(args);
-
-      const parent = (segment instanceof Subsegment
-        ? segment.segment
-        : segment);
-
-      args.request.headers['X-Amzn-Trace-Id'] = stringify(
-        {
-          Root: parent.trace_id,
-          Parent: segment.id,
-          Sampled: parent.notTraced ? '0' : '1',
-        },
-        ';',
-      );
-      return next(args);
-    },
-    {
-      step: 'build',
-    },
-  );
 
+    const errObj = { message: err.message, name: err.name, stack: err.stack || new Error().stack };
+    subsegment.close(errObj, true);
+    throw err;
```

Comment:
Is throwing an Object instead of Error expected in Xray SDK?

Reply:
What do you mean? We construct the `errObj` to record it in the X-Ray subsegment, so it conforms to our data model. But after recording it we just throw the original caught error.

Reply:
My bad, I mistakenly read as errObj being thrown. There's no issue here.